src/concurrency.h

```diff
@@ -8,140 +8,178 @@
 #include <thread>
 #include <atomic>
 #include <functional>
+#include <cassert>
 
 const int QUEUE_MAX_SIZE = 10;
 // Simplest implementation of a blocking concurrent queue for thread messaging
 template<class T>
 class single_consumer_queue
 {
     std::deque<T> _queue;
-    std::mutex _mutex;
+    mutable std::mutex _mutex;
     std::condition_variable _deq_cv; // not empty signal
-    std::condition_variable _enq_cv; // not empty signal
+    std::condition_variable _enq_cv; // not full signal
 
-    unsigned int _cap;
+    unsigned int const _cap;
     bool _accepting;
 
-    // flush mechanism is required to abort wait on cv
-    // when need to stop
-    std::atomic<bool> _need_to_flush;
-    std::atomic<bool> _was_flushed;
-    std::function<void(T const &)> _on_drop_callback;
+    std::function<void(T const &)> const _on_drop_callback;
+
 public:
-    explicit single_consumer_queue<T>(unsigned int cap = QUEUE_MAX_SIZE, std::function<void(T const &)> on_drop_callback = nullptr)
-        : _queue(), _mutex(), _deq_cv(), _enq_cv(), _cap(cap), _accepting(true), _need_to_flush(false), _was_flushed(false), _on_drop_callback(on_drop_callback)
-    {}
+    explicit single_consumer_queue< T >( unsigned int cap = QUEUE_MAX_SIZE,
+                                         std::function< void( T const & ) > on_drop_callback = nullptr )
+        : _cap( cap )
+        , _accepting( true )
+        , _on_drop_callback( on_drop_callback )
+    {
+    }
 
+    // Enqueue an item onto the queue.
+    // If the queue grows beyond capacity, the front will be removed, losing whatever was there!
     void enqueue(T&& item)
     {
         std::unique_lock<std::mutex> lock(_mutex);
-        if (_accepting)
+        if( ! _accepting )
         {
-            _queue.push_back(std::move(item));
-            if (_queue.size() > _cap)
-            {
-                if (_on_drop_callback)
-                {
-                    _on_drop_callback(_queue.front());
-                }
-                _queue.pop_front();
-            }
+            if( _on_drop_callback )
+                _on_drop_callback( item );
+            return;
         }
+
+        _queue.push_back(std::move(item));
+
+        if( _queue.size() > _cap )
+        {
+            if( _on_drop_callback )
+                _on_drop_callback( _queue.front() );
+            _queue.pop_front();
+        }
+
         lock.unlock();
+
+        // We pushed something -- let others know there's something to dequeue
         _deq_cv.notify_one();
     }
 
-    void blocking_enqueue(T&& item)
-    {
-        auto pred = [this]()->bool { return _queue.size() < _cap || _need_to_flush; };
 
+    // Enqueue an item, but wait for room if there isn't any
+    // Returns true if the enqueue succeeded
+    bool blocking_enqueue(T&& item)
+    {
         std::unique_lock<std::mutex> lock(_mutex);
-        if (_accepting)
+        _enq_cv.wait( lock, [this]() {
```

Comment:
```
wait( lock, pred )
```
is the same as:
```
while( ! pred() ) {
    wait( lock );
}

Reply:
Is we are not accepting, the queue should be empty, so we shouldn't really wait, no?

Reply:
So you are relying on logic elsewhere that say (accepting = false) is the same as (size = 0)

Reply:
Well yes, I guess so: compound predicates get less readable because of the `!`. Would it be better if I wrote the pred as `return _accepting  &&  _queue.size() < _cap;` ?

Reply:
Actually, it gets more complicated...
You want:
```
while( _accepting && _queue.size() >= _cap ) {
    wait();
}
```
So the logic inside the pred should be `return ! _accepting || _queue.size() < _cap;`...